[RF] Migrate HistFactory from legacy iterators to range-based loops

subham73 opened the pull request:
# This Pull request:

## Changes or fixes: Migration from RooAbsCollection and RooLinkedList legacy iterators to range-based loops
PR includes migration of all the flies from Roofit/histfactory/src
## Checklist:

- [x] tested changes locally
- [ ] updated the docs (if necessary)

This PR fixes #8777

guitargeek commented on roofit/histfactory/src/FlexibleInterpVar.cxx:
```diff
@@ -86,14 +86,12 @@ FlexibleInterpVar::FlexibleInterpVar(const char* name, const char* title,
 
-  RooFIter highIter = highList.fwdIterator() ;
-  while ((val = (RooAbsReal*) highIter.next())) {
+  //RooFIter highIter = highList.fwdIterator() ;
```
```suggestion
```
This commented line should be deleted

guitargeek commented on roofit/histfactory/src/HistFactoryModelUtils.cxx:
```diff
@@ -185,11 +182,8 @@ namespace HistFactory{
     bool FoundParamHistFunc=false;
-    RooAbsArg* paramfunc_arg = NULL;
-    while(( paramfunc_arg = (RooAbsArg*) iter->Next() )) {
+    for (auto const *paramfunc_arg : static_range_cast<RooAbsArg *>(*channel->getComponents())) {
```
```suggestion
    for (auto const *paramfunc_arg : *channel->getComponents()) {
```
RooAbsArg is already the class of the components (RooAbsArg is the base class of the other RooFit object classes), so this cast is unnecessary.

guitargeek commented on roofit/histfactory/src/HistFactoryModelUtils.cxx:
```diff
@@ -226,11 +217,9 @@ namespace HistFactory{
     // get category label
-    RooArgSet* allobs = (RooArgSet*) data->get();
-    TIterator* obsIter = allobs->createIterator();
+    RooArgSet *allobs = (RooArgSet *)data->get();
```
```suggestion
    RooArgSet *allobs = data->get();
```
`RooAbsData::get()` already returns a `RooArgSet*`, no cast is necessary.

guitargeek commented on roofit/histfactory/src/HistFactoryModelUtils.cxx:
```diff
@@ -311,12 +299,10 @@ namespace HistFactory{
     bool FoundConstraintTerm=false;
     RooAbsPdf* constraintTerm=NULL;
-    while((term_constr=(RooAbsArg*)iter_list->Next())) {
+    for (auto const *term_constr : static_range_cast<RooAbsArg *>(*constraints)) {
```
```suggestion
    for (auto const *term_constr : *constraints) {
```

guitargeek commented on roofit/histfactory/src/HistFactoryNavigation.cxx:
```diff
@@ -753,25 +736,20 @@ namespace RooStats {
 
       // Get All Subnodes of this product
-      RooArgSet productComponents = node->components();
+      //RooArgSet productComponents = node->components();
```
```suggestion
```
This commented line can be removed.

guitargeek commented on roofit/histfactory/src/HistFactoryNavigation.cxx:
```diff
@@ -753,25 +736,20 @@ namespace RooStats {
-   else {
-     allTerms.add(*arg);
-   }
+      for (auto *arg : dynamic_range_cast<RooAbsArg *>(node->components())) {
```
```suggestion
      for (auto *arg : node->components()) {
```
Avoid unnecessary cast.

guitargeek commented on roofit/histfactory/src/HistFactoryNavigation.cxx:
```diff
@@ -753,25 +736,20 @@ namespace RooStats {
+      for (auto *arg : dynamic_range_cast<RooAbsArg *>(node->components())) {
+        std::string ClassName = arg->ClassName();
+        if( ClassName == "RooProduct" ) {
+            RooProduct* prod = dynamic_cast<RooProduct*>(arg);
```
```suggestion
            RooProduct* prod = static_cast<RooProduct*>(arg);
```
Can you change this to `static_cast`? The previous line already makes sure that this is a `RooProduct`.

guitargeek commented on roofit/histfactory/src/HistFactoryNavigation.cxx:
```diff
@@ -839,16 +817,14 @@ namespace RooStats {
-       break;
-     }
-   }
+    for (auto const *arg : static_range_cast<RooAbsArg *>(*components)) {
```
```suggestion
    for (auto const *arg : *pdf->getComponents()) {
```
Unnecessary cast, and I guess you can collapse this with the previous line that can be removed?

guitargeek commented on roofit/histfactory/src/HistFactoryNavigation.cxx:
```diff
@@ -839,16 +817,14 @@ namespace RooStats {
+    for (auto const *arg : static_range_cast<RooAbsArg *>(*components)) {
+      std::string ClassName = arg->ClassName();
+      if( ClassName == "RooRealSumPdf" ) {
+      fChannelSumNodeMap[ChannelName] = (RooRealSumPdf*) arg;
```
```suggestion
      fChannelSumNodeMap[ChannelName] = static_cast<RooRealSumPdf*>(arg);
```
If you touch a line that uses these C-style casts, you should update them to C++ casts while at it.

guitargeek commented on roofit/histfactory/src/HistFactoryNavigation.cxx:
```diff
@@ -871,10 +847,10 @@ namespace RooStats {
-   RooAbsArg* sample;
-   while( (sample=(RooAbsArg*)sampleItr.Next()) ) {
-
+   //TIter sampleItr = nodes.createIterator();
+   //RooAbsArg* sample;
+   //while( (sample=(RooAbsArg*)sampleItr.Next()) ) 
+   for (auto const *sample : static_range_cast<RooAbsArg *>(nodes)) {      
```
```suggestion
   for (auto * func : static_range_cast<RooAbsReal*>(sumPdf->funcList())) {
```
Here you can do more things in one line: you can already cast to `RooAbsReal*` here to absorb the line below, and replace `sample` with `func` in the code below. You also don't need the temporary `nodes` and the old commented-out lines can be removed.

guitargeek commented on roofit/histfactory/src/HistFactoryNavigation.cxx:
```diff
@@ -910,38 +886,30 @@ namespace RooStats {
-     term = arg; //dynamic_cast<RooAbsReal*>(arg);
-     break;
-   }
+      for (auto arg : static_range_cast<RooAbsArg *>(*parent->getComponents())) {
```
```suggestion
      for (auto arg : *parent->getComponents()) {
```

guitargeek commented on roofit/histfactory/src/HistFactoryNavigation.cxx:
```diff
@@ -910,38 +886,30 @@ namespace RooStats {
-     term = param; //dynamic_cast<RooAbsReal*>(arg);
-     break;
-   }
+      RooArgSet* paramSet = parent->getParameters(args); 
+      for (auto *param : static_range_cast<RooAbsArg *>(*paramSet)) {
```
```suggestion
      for (auto *param : *parent->getParameters(args)) {
```
Remove unnecessary cast, and I think the temporary `paramSet` can be avoided.

guitargeek commented on roofit/histfactory/src/HistFactoryNavigation.cxx:
```diff
@@ -1127,9 +1095,10 @@ namespace RooStats {
-   TIter itr = components.createIterator();
-   RooAbsArg* arg = NULL;
-   while( (arg=(RooAbsArg*)itr.Next()) ) {
+   //TIter itr = components.createIterator();
+   //RooAbsArg* arg = NULL;
+   //while( (arg=(RooAbsArg*)itr.Next()) ) 
+   for (auto *arg : dynamic_range_cast<RooAbsArg *>(components)) {
```
```suggestion
   for (auto * component : static_range_cast<RooAbsReal *>(components)) {
```
Remove commented out lines, and you can directly cast to `RooAbsReal` which is the type required in this loop (you can delete the line below then). Also it can be a `static_cast` because the result is never checked for `nullptr` anyway.

guitargeek commented on roofit/histfactory/src/HistFactoryNavigation.cxx:
```diff
@@ -1150,32 +1119,31 @@ namespace RooStats {
-   PrintMultiDimHist(hist, bin_print_width);
-   delete hist;
+ 
+      for (auto *arg : static_range_cast<RooAbsArg *>(components)) {
+        RooAbsReal* component = dynamic_cast<RooAbsReal*>(arg);
```
```suggestion
      for (auto *component : static_range_cast<RooAbsReal *>(components)) {
```
Again, you can directly cast down to `RooAbsReal`.

guitargeek commented on roofit/histfactory/src/HistoToWorkspaceFactory.cxx:
```diff
@@ -998,13 +994,11 @@ namespace HistFactory{
-    TIterator* params_itr=POIs->createIterator();
-    TObject* params_obj=0;
-    while((params_obj=params_itr->Next())){
+    for (auto const *params_obj : static_range_cast<RooRealVar *>(*POIs)) {
```
```suggestion
    for (auto const *poi : static_range_cast<RooRealVar *>(*POIs)) {
```
Look at the line below that can now deleted, you are redundantly casting to RooRealVar.

guitargeek commented on roofit/histfactory/src/MakeModelAndMeasurementsFast.cxx:
```diff
@@ -374,12 +374,7 @@ void RooStats::HistFactory::FitModelAndPlot(const std::string& MeasurementName,
-  while( (poi_obj=params_itr.Next()) ) {
-    //poi = (RooRealVar*) poi_obj;
-    poi = dynamic_cast<RooRealVar*>(poi_obj);
+  for (auto const *poi : dynamic_range_cast<RooRealVar *>(*POIs)) {
```
```suggestion
  for (auto const *poi : static_range_cast<RooRealVar *>(*POIs)) {
```
We can use static cast here because the output is never checked for `nullptr`.

guitargeek commented on roofit/histfactory/src/MakeModelAndMeasurementsFast.cxx:
```diff
@@ -388,7 +383,7 @@ void RooStats::HistFactory::FitModelAndPlot(const std::string& MeasurementName,
   // But we only make detailed plots and tables
   // for the 'first' POI
-  poi = dynamic_cast<RooRealVar*>(POIs->first());
+  RooRealVar* poi = dynamic_cast<RooRealVar *>(POIs->first());
```
```suggestion
  RooRealVar* poi = static_cast<RooRealVar *>(POIs->first());
```

guitargeek commented on roofit/histfactory/src/PiecewiseInterpolation.cxx:
```diff
@@ -87,9 +87,7 @@ PiecewiseInterpolation::PiecewiseInterpolation(const char* name, const char* tit
-  RooFIter inputIter1 = lowSet.fwdIterator() ;
-  RooAbsArg* comp ;
-  while((comp = inputIter1.next())) {
+  for (auto *comp : static_range_cast<RooAbsArg *>(lowSet)) {
```
```suggestion
  for (auto *comp : lowSet) {
```

guitargeek commented on roofit/histfactory/src/PiecewiseInterpolation.cxx:
```diff
@@ -102,8 +100,7 @@ PiecewiseInterpolation::PiecewiseInterpolation(const char* name, const char* tit
 
-  RooFIter inputIter2 = highSet.fwdIterator() ;
-  while((comp = inputIter2.next())) {
+  for (auto *comp : static_range_cast<RooAbsArg *>(highSet)) {
```
```suggestion
  for (auto *comp : highSet) {
```

guitargeek commented on roofit/histfactory/src/PiecewiseInterpolation.cxx:
```diff
@@ -116,8 +113,7 @@ PiecewiseInterpolation::PiecewiseInterpolation(const char* name, const char* tit
 
-  RooFIter inputIter3 = paramSet.fwdIterator() ;
-  while((comp = inputIter3.next())) {
+  for (auto *comp : static_range_cast<RooAbsArg *>(paramSet)) {
```
```suggestion
  for (auto *comp : paramSet) {
```

subham73 commented:
Taking note of all the things will try to modify them as suggested and look again in all files and will surely avoid casting when it can be avoided . Will PR this above files and new files after that :). Thank you for your time and suggestions.

guitargeek reviewed (changes requested):
Hi @subham73, thanks for opening the PR and helping out, this is really great that you managed to compile ROOT and make your first contribution!

I have made some change requests, that address mostly 3 points:
* casting to `RooAbsArg*` is not necessary
* some `dynamic_cast` can be replaced with `static_cast`
* C-style casts should be avoided if you already touch the line of code

These issues should be addressed before the PR can be merged. Just add commits to this branch, it's no problem for GitHub to update the branch while the PR is open :)

Thanks again for your help, this is a great service to ROOT and RooFit :D

guitargeek commented:
@phsft-bot build

guitargeek commented:
Hi @subham73, thank you very much for updating the PR! I have started a first run of tests.

If the tests pass and after you addressed the remaining comments of mine about using `static_range_cast` instead of `dynamic_range_cast`, we can merge this one.

subham73 commented:
> Hi @subham73, thank you very much for updating the PR! I have started a first run of tests.
> 
> If the tests pass and after you addressed the remaining comments of mine about using `static_range_cast` instead of `dynamic_range_cast`, we can merge this one.

Hello @guitargeek !! At that time I only updated 2 of the files, now I have modified all the rest files of histfactory/src. 
Can you pls provide the review on src/LinInterpVar.cxx, should I just leave the instance of the constructor like that.
Also i don't really know why this test [`[projectroot.roottest.python.distrdf.spark.roottest_python_distrdf_spark_spark_test_rungraphs]`]((https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/138804/testReport/projectroot.roottest.python.distrdf/spark/roottest_python_distrdf_spark_spark_test_rungraphs/)) is falling . Thankyou for ur time and also for any new suggestions and corrections before merging.

guitargeek commented on roofit/histfactory/src/LinInterpVar.cxx:
```diff
@@ -100,7 +93,7 @@ LinInterpVar::LinInterpVar(const LinInterpVar& other, const char* name) :
 LinInterpVar::~LinInterpVar()
 {
-  delete _paramIter ;
+  //delete _paramIter ;
```
```suggestion
```
This commented-out line can be deleted. Actually, the `_paramIter` member is not used anymore now, great!
That means you should remove it also from the class members in `roofit/histfactory/inc/RooStats/HistFactory/LinInterpVar.h`.

Also, the destructor `~LinInterpVar()` does nothing now, so you can remove it as well (also from the header file).

guitargeek commented:
Thanks for updating! Your changes to `src/LinInterpVar.cxx` are correct, you were just missing the corresponding changes in the header file. And the now-empty destructor can be removed.

guitargeek commented:
@phsft-bot build

guitargeek commented:
@phsft-bot build

guitargeek reviewed (approved):
LGTM, thank you so much @subham73 for helping out with this! I hope you continue to contribute to RooFit, this is really a great service :+1:
